contrib/parseq-zk-client/src/main/java/com/linkedin/parseq/zk/client/ZKClient.java

```diff
@@ -302,8 +304,34 @@ protected Promise run(Context context) throws Throwable {
    * @param executor {@code Executor} that will be used to run the operations.
    * @return task to execute multiple operations.
    */
-  public Task<List<OpResult>> multi(Iterable<Op> ops, Executor executor) {
-    return Task.blocking(() -> _zkClient.multi(ops), executor);
+  public Task<List<OpResult>> multi(List<Op> ops, Executor executor) {
+    return Task.blocking(() -> _zkClient.multi(ops), executor)
+        .recoverWith(e -> {
+          if (e instanceof KeeperException) {
+            e = getCause((KeeperException) e, ops);
+          }
+          return Task.failure(e);
+        });
+  }
+
+  /**
+   * Gets the exact cause out of {@link KeeperException} thrown from
+   * {@link ZooKeeper#multi(Iterable)}
+   *
+   * @param ex {@code KeeperException}
+   * @param ops list of {@code Op}s which resulted in the given {@code KeeperException}.
+   * @return unwrapped {@code KeeperException} with the correct error code and path info.
+   */
+  private KeeperException getCause(KeeperException ex, List<Op> ops) {
+    List<OpResult> results = ex.getResults();
+    for (int i = 0; i < results.size(); ++i) {
+      if (results.get(i) instanceof OpResult.ErrorResult &&
+          ((OpResult.ErrorResult) results.get(i)).getErr() != KeeperException.Code.OK.intValue()) {
+        OpResult.ErrorResult error = (OpResult.ErrorResult) results.get(i);
+        return KeeperException.create(KeeperException.Code.get(error.getErr()), ops.get(i).getPath());
+      }
+    }
+    throw new IllegalArgumentException("None of the Ops is failed. Ops results: " + ex.getResults());
```

Comment:
In theory, we should never reach this line. So I'd rather throw exception here to (hopefully) raise more attention.